### DISCUSSION TITLE
How to disable immediate checks for Unhandled Promise rejections?

### DISCUSSION THREAD

**Discussion by jcubic** [avajs/ava#3289] [Q&A] [answered]
I have async tests I'm testing complex async code (programming language) and I'm just working on try..catch I've added code that throws an error inside the Promise constructor that is caught. THe code works fine when I try to replicate the same code from unit tests in NodeJS but Ava keeps showing me an error:

```
  Unhandled rejection in tests/test.js

  Error: <CATCH>

  › file://src/lips.js:8973:43
  › unpromise (file://src/lips.js:1730:12)
  › error (file://src/lips.js:8965:29)
```

I was reading the code and it seems that ava uses `currently-unhandled` which checks if the promise is not immediately handled, is there a way to disable this "feature" I have complex code, and all promises are handled inside the code but I may take longer then one tick to handle the rejection.

There are no unhandled rejections in my code.

To explain my use case. I had a programming language (Scheme) and I've had a problem with `try..catch`

```
(try
  (begin
   (print "<X>")
   (throw (new Error "<ERR>"))
   (print "<Y>"))
  (catch (e)
    (print "catch")))
```

The problem was the `(print "<Y>")` was executed even though the error was thrown. I fixed it by throwing fake exception (inside `try`) and handling it on some higher level actually the exception is caught and ignored. But because my code has multiple levels or async code. This does not guarantee that the handle of this exception will take place inside the same tick. The promise rejection never propagates outside of the test function.

So I need to disable this flaky unhandled promise rejection. Even if this is for a single test.

There should be a way to have some hooks inside the library to change it. Unfortunately, I can't use any other library since this is the only one that is async and works with my code.

**EDIT**: I was trying to reproduce the issue with a smaller code, but I was not able to do this.

**Comment by jcubic** [comment, marked as answer]
I was able to fix the issue by ignoring the exception I used to escape from try..catch:

```javascript
    process.on('unhandledRejection', (reason, promise) => {
        if (reason instanceof IgnoreException) {
            promise.catch(noop);
        }
    });
```

The same function can be used to ignore any unhandled rejection reported by AVA.

**Comment by novemberborn** [comment]
I believe `currenty-handled` discards rejections that have been handled since.

In other words: it should only highlight rejected promises that are unhandled when AVA has completed the test execution. I do think that is legitimate.
